mlir/lib/Dialect/Rock/IR/RockDialect.cpp

```diff
@@ -1435,16 +1435,35 @@ LogicalResult ThreadwiseReadIntoOp::verify() {
   auto gpuMemSpaceAttr = memSpaceAttr.dyn_cast_or_null<gpu::AddressSpaceAttr>();
   if (memSpaceAttr && (!gpuMemSpaceAttr || gpuMemSpaceAttr.getValue() !=
                                                gpu::AddressSpace::Private))
-    return emitOpError("source must be private registers");
+    return emitOpError("dest must be private registers");
   ArrayAttr extraViews = getExtraViews();
   ArrayRef<int64_t> inputShape;
   if (extraViews.empty())
     inputShape = getSource().getType().getShape();
   else
     inputShape = extraViews[0].cast<TransformMapAttr>().getUpperBounds();
 
-  if (inputShape.size() != 3)
-    return emitOpError("source view must accept (bid, tid, iter) coordinates");
+  MemRefType srcType = getSource().getType().cast<MemRefType>();
+  Attribute srcMemSpaceAttr = srcType.getMemorySpace();
+  // Unless specified it is assumed to be global
+  gpu::AddressSpace srcGpuMemSpace = gpu::AddressSpace::Global;
+  if (srcMemSpaceAttr) {
+    srcGpuMemSpace = srcMemSpaceAttr.cast<gpu::AddressSpaceAttr>().getValue();
```

Comment:
`auto srcMemSpaceAttr = srcType.getMemorySpace.dyn_cast_or_null<gpu::AddressSpaceAttr>();`
and then the if statement. Or possibly erroring out on memory spaces that aren't either absent or the GPU ones.
That cast<> will cause crashes on user input as opposed to programmer error.

Reply:
good call! done.